src/main/java/io/anserini/document/JsonDocument.java

```diff
@@ -34,45 +36,38 @@
 public class JsonDocument implements SourceDocument {
   protected String id;
   protected String contents;
-  private JsonElement raw;
+  private ArrayNode node;
   private int i;
 
-  public JsonDocument(String path) throws FileNotFoundException {
-    JsonParser parser = new JsonParser();
+  public JsonDocument(String path) {
     try {
-      raw = parser.parse(new BufferedReader(new FileReader(path))).getAsJsonArray();
-    } catch (IllegalStateException e1) {
-      try {
-        raw = parser.parse(new BufferedReader(new FileReader(path))).getAsJsonObject();
-      } catch (IllegalStateException e2) {
-        raw = parser.parse(new BufferedReader(new FileReader(path))).getAsJsonNull();
-      }
+      JsonParser jsonParser = new JsonFactory().createParser(new BufferedReader(new FileReader(path)));
+      ObjectMapper objectMapper = new ObjectMapper();
+      objectMapper.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+      node = objectMapper.readTree(jsonParser);
+    } catch (IOException e) {
+      node = null;
     }
     i = 0;
   }
 
   @Override
   public JsonDocument readNextRecord(BufferedReader bRdr) throws IOException {
-    if (raw.isJsonArray() && i < raw.getAsJsonArray().size()) {
-      JsonObject o = raw.getAsJsonArray().get(i).getAsJsonObject();
-      id = o.get("id").getAsString();
-      contents = o.get("contents").getAsString();
-      i++;
-      return this;
-    } else if (raw.isJsonObject()) {
-      id = raw.getAsJsonObject().get("id").getAsString();
-      contents = raw.getAsJsonObject().get("contents").getAsString();
-      return this;
-    } else {
+    if (node == null || i >= node.size()) {
```

Comment:
Under that condition can `i >= node.size()`?

Reply:
when we have already read and parsed everything in `node` (i.e. `raw` in old code)

Reply:
So either an array of objects or a single object will be stored in `ArrayNode node`, right?
And that will be handled in the below `else` block. 
My understanding of `i >= node.size()` is that it will only be executed when there is a mixed array and line objects file, no?

Reply:
Yes, either an array of JSON objects or a single JSON object will be stored in `ArrayNode node`.

When we have already read and parsed all JSON objects in `node`, we will read and parse from `BufferedReader bRdr`. This is the same logic as `i < raw.getAsJsonArray().size()` in the old code, right?

Reply:
In the old code, we first categorize the different types:  `raw.isJsonArray()` or `raw.isJsonObject()` or the rest (line json object). There can not be mixed types.
Here, `node==null` means the file contains only line json objects and should be treated as such. So we will never hit `i >= node.size()`.

Reply:
When `JsonDocument` is created, will the JSON file contain a mixed content of both array of JSON objects and individual JSON objects?

Reply:
I see. So `i >= node.size()` is unnecessary. I'll remove it.

Reply:
Yea, let's assume that the file is either:

1. An array of json objects (valid JSON format)
2. A single json object (valid JSON format)
3. line json objects (invalid JSON format, only be used for tweet-like files)

Reply:
Agreed, but the `i >= node.size()` should be in yet another `else if` branch, right?